pkgpanda/util.py

```diff
@@ -264,13 +265,32 @@ def make_file(name):
 
 
 def write_json(filename, data):
-    with open(filename, "w+") as f:
-        return json.dump(data, f, **json_prettyprint_args)
+    dumped_json = json.dumps(data, **json_prettyprint_args)
+    write_string(filename, dumped_json)
 
 
 def write_string(filename, data):
-    with open(filename, "w+") as f:
-        return f.write(data)
+    """
+    Write a string to a file.
+    Overwrite any data in that file.
+
+    We use an atomic write practice of creating a temporary file and then
+    moving that temporary file to the given ``filename``. This prevents race
+    conditions such as the file being read by another process after it is
+    opened here but not yet written to.
+
+    It also prevents us from creating or truncating a file before we fail to
+    write data to it because of low disk space.
+    """
+    temporary_filename = '{filename}-{random}.tmp'.format(
+        filename=filename,
+        random=uuid.uuid4().hex,
+    )
+
+    with open(temporary_filename, 'w') as f:
```

Comment:
From creation of the temporary file until the `os.replace` needs exception handling to remove the file if an error occurs, especially since you appear to be worried about disk full conditions. But it's not nice to have random files appearing of unknown significance either.

This can be a bit fiddly with the current code, since you create the file in a context manager. You'd have to start the `try` before the `open`, so the exception handling has to check if the file exists before removing.

With `mkstemp` the file exists when the `mkstemp` call exits, so the code can unconditionally remove the file. Note, it needs a double-try, one to close the file descriptor and one to remove the file on exception.